### ISSUE TITLE
fluentd runs on every node; we should find something that is more efficient

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#23782]
We should investigate more efficient alternatives to fluentd, because it runs on every node, so effectively takes 10% of your entire cluster.

**Comment by edsiper** [comment]
would you please elaborate more on this issue ?

**Comment by justinsb** [comment]
@edsiper sure - this was just one of a number of issues that I filed while trying to figure out why a user's new cluster didn't come up, because the k8s system by default was using more than 2 m3.mediums for its own purposes!  (Mostly because of kube-proxy's allocation, but that is fixed in 1.2.1)

We run fluentd on every node in k8s, and we allocate it 0.1 of a core, although I think we're changing in to 0.08.  A user has pointed out this allocation is a pretty high tax to pay.  So we should find a way to make fluentd more lightweight, or - probably - investigate something that is compiled.

There are also a number of other shortcomings of the current log collection approach that I would love to see solved:

1. We need to integrate with AWS hosted ES (although a plugin is available)
1. It would be nice to collect labels / annotations from the pod
1. Users have reported problems parsing multi-line entries (e.g. Java), I guess because we have to decode the docker entries and _then_ apply the multi-line parsing

**Comment by therc** [comment]
Or log collection could be deferred to the Docker side, assuming we can still collect labels.

**Comment by therc** [comment]
Also, hosted AWS ES is busted because of this: https://discuss.elastic.co/t/elasitcsearch-ruby-raises-cannot-get-new-connection-from-pool-error/36252/11

**Comment by thockin** [comment]
There is and will always be a tension between doing work on each node
(which scales naturally and provides good latency and availability
characteristics) and using a little overhead as possible (ie. centralizing
work).

Even inside Google this is an ongoing issue.  The truth is that software
tends to grow over time.  Whether that is new features of existing programs
or new programs.  For physical clusters we can get away with it somewhat by
saying that machine sizes grow faster than we do (the only 4-core machine
you can buy is a cell phone), so we are a diminishing percentage of the
machine.  This is less true for VMs, where smaller VMs offer some
flexibility advantages.

There are already plans/hopes for moving things like DNS onto each node.

The TL;DR is that it quickly becomes a rock-vs-hardplace situation.  You
can either stop enhancing the system or you have to allow for more overhead
over time.  In the mean time we can try to minimize the impact of each
per-node agent and of the per-cluster agents, but we need to not pretend
that overhead doesn't exist.

**Comment by tagomoris** [comment]
From my viewpoint, some of these problems are brought by using Fluentd + in_tail + json container logs.
This is because Kubernetes's use of Fluentd predates Fluentd's Docker Logging Driver. As a result, it requires a container (and some CPU resource) per host at least.

If Kubernetes used Docker Logging Driver with Fluentd, the user can configure Fluentd in a couple of ways:

* logging driver -> localhost fluentd (per-host aggregator) -> cluster aggregator (ES? Fluentd+something?)
  * per-host aggregator provides network connectivity & buffering for containers in same host
* logging driver -> Fluentd (cluster aggregator) -> ES or something
  * this configuration requires just a Fluentd node per cluster (if high logging throughput is NOT required)

In addition, Fluentd logging driver provides labels as an event/record field.

My guess is Kubernetes uses Fluentd's multiline parser to handle multiline logs.
Unfortunately, Fluentd + docker logging driver solution still has problem about multiline logs because Docker logging driver splits a log with newlines into
multiple lines of logs. It requires some more development on Docker's logging driver mechanism itself and logging driver for Fluentd as well as Fluentd plugins.

**Comment by bgrant0607** [comment]
@tagomoris By "labels", do you mean Docker labels, or arbitrary key-value pairs, or something else?

**Comment by bgrant0607** [comment]
@tagomoris We also need to be able to handle logging to files, which I assume the Docker logging driver does not.

**Comment by tagomoris** [comment]
@bgrant0607 Fluentd logging driver can handle `ExtraAttributes` from logger context of docker, and it was described as (in daemon/logger/context.go):
```
// ExtraAttributes returns the user-defined extra attributes (labels,
// environment variables) in key-value format. This can be used by log drivers
// that support metadata to add more context to a log.
```

**Comment by piosz** [comment]
We are working with fluentd guys to limit memory consumption of fluentd right now. See #32762 for more details. Hopefully we will be able to do it. Otherwise we will revisit this issue.

In the meantime closing in favor of #32762.

**Comment by piosz** [comment]
cc @Crassirostris
